### PR TITLE
monty-31: improve `monty_reduce_u128` speed

### PR DESCRIPTION
**Pull request by tcoratger**
I feel it doesn't make much sense to include it but I've setup the following meaningful benchmark file for the dot product which uses `monty_reduce_u128` under the hood. Here is the used benchmark file:

```rust
//! Benchmarks exercising `monty_reduce_u128` via `dot_product`.
//!
//! For N > 8, `MontyField31::dot_product` accumulates into a `u128` and
//! calls `monty_reduce_u128` for the final reduction. Comparing different
//! sizes shows the impact of the reduction strategy.

use criterion::{black_box, criterion_group, criterion_main, BenchmarkId, Criterion};
use p3_field::PrimeCharacteristicRing;
use p3_koala_bear::KoalaBear;
use rand::rngs::SmallRng;
use rand::{Rng, RngExt, SeedableRng};

type F = KoalaBear;

fn bench_dot_product(c: &mut Criterion) {
    let mut group = c.benchmark_group("dot_product_koalabear");

    macro_rules! bench_dp {
        ($n:literal) => {{
            group.bench_with_input(BenchmarkId::new("N", $n), &$n, |b, _| {
                let mut rng = SmallRng::seed_from_u64(123);
                let a: [F; $n] = core::array::from_fn(|_| rng.random());
                let b_arr: [F; $n] = core::array::from_fn(|_| rng.random());
                b.iter(|| {
                    black_box(F::dot_product::<$n>(black_box(&a), black_box(&b_arr)))
                });
            });
        }};
    }

    // N <= 8: uses u64 path (large_monty_reduce), baseline for comparison
    bench_dp!(4);
    bench_dp!(8);
    // N > 8: uses u128 accumulation + monty_reduce_u128
    bench_dp!(16);
    bench_dp!(64);
    bench_dp!(256);
    bench_dp!(1024);

    group.finish();
}

criterion_group!(benches, bench_dot_product);
criterion_main!(benches);
```

Here are the obtained results after this PR:
Capture d’écran 2026-03-25 à 19 09 56

### CODE REVIEW

File: monty-31/src/utils.rs [hunk 1]
```diff
@@ -157,22 +157,438 @@ pub(crate) const fn large_monty_reduce<MP: MontyParameters>(x: u64) -> u32 {
...
+    let hi_r_mod_p = to_monty::<MP>(hi);
+
+    // Combine the two reduced halves with a modular addition.
+    //
+    // Both operands are in [0, P), so their sum is in [0, 2*P).
+    // A single conditional subtraction of P yields a result in [0, P).
+    add::<MP>(hi_r_mod_p, r)
+}
+
+#[cfg(test)]
```
Comment (tcoratger) [inline monty-31/src/utils.rs:224]: I've added some proptests here because this utils module was not tested at all and as it is the heart of many operations and some edge cases can go unnoticed, I imagined that it was good to have some small tests (especially when we are doing PRs like this to double confirm that we are not doing something stupid).

But this is totally unrelated with the performance improvement. So let me know if you feel this is not needed and I can revert this part.

File: monty-31/src/utils.rs [hunk 2]
```diff
@@ -157,22 +157,438 @@ pub(crate) const fn large_monty_reduce<MP: MontyParameters>(x: u64) -> u32 {
...
+///
+/// # Performance
+///
+/// All arithmetic stays at 64 bits or below:
+/// - One Montgomery reduction on a `u64` (multiplies and shifts, no division).
+/// - One `u64 % P` where `P` is a compile-time constant (LLVM emits a
+///   multiply-and-shift sequence, ~3-5 cycles).
+/// - One conditional subtraction for the final modular addition.
+///
+/// No 128-bit division or modulo is ever performed.
```
Comment (Nashtare) [inline monty-31/src/utils.rs]: not necessary given section above
```suggestion
```

Comment (Nashtare) [inline monty-31/src/utils.rs]: I'm not sure we need this whole section given that we re-describe the algorithm within the function itself?

File: monty-31/src/utils.rs (hunk 1 above)
Comment (Nashtare) [reply monty-31/src/utils.rs:224]: Happy to bring in more tests tbh

File: monty-31/src/utils.rs [hunk 3]
```diff
@@ -157,22 +157,438 @@ pub(crate) const fn large_monty_reduce<MP: MontyParameters>(x: u64) -> u32 {
...
-/// Perform a monty reduction on a u128 in the range `[0, 2^96)`
+/// Montgomery reduction of a 128-bit value, returning `x * R^{-1} mod P`.
 ///
-/// The input will be in `[0, P)` and be equal to `x * MONTY^{-1} mod P`.
+/// where `R = 2^MONTY_BITS = 2^32`.
+///
+/// # Safety
+///
+/// - Input must satisfy `x < 2^96`.
+/// - Output is in `[0, P)`.
```
Comment (Nashtare) [inline monty-31/src/utils.rs]: `# Safety` is usually used for `unsafe` functions per Rust standards so it is a bit confusing here.

Comment (Nashtare) [reply monty-31/src/utils.rs]: also we weren't doing it before but I think it could be good to have a `debug_assert!(x < 1 << 96)` inside?

Comment (Nashtare) [reply monty-31/src/utils.rs]: also makes me realize we probably should have a const assert for MONTY_BITS to be 32 right? As the function technically accepts wider types (even though it outputs a u32).

File: monty-31/src/utils.rs [hunk 4]
```diff
@@ -157,22 +157,438 @@ pub(crate) const fn large_monty_reduce<MP: MontyParameters>(x: u64) -> u32 {
...
+/// - The low limb `lo * R^{-1} mod P` is a standard Montgomery reduction on a `u64`.
+/// - The high limb `hi * 2^32 mod P` is a conversion into Montgomery form.
+/// - The two residues are combined with a single modular addition.
+///
+/// # Performance
+///
+/// All arithmetic stays at 64 bits or below:
+/// - One Montgomery reduction on a `u64` (multiplies and shifts, no division).
+/// - One `u64 % P` where `P` is a compile-time constant (LLVM emits a
+///   multiply-and-shift sequence, ~3-5 cycles).
```
Comment (Nashtare) [inline monty-31/src/utils.rs]:
```suggestion
/// - One `u64 % P` where `P` is a compile-time constant.
```

simply because this is hard to test/measure/enforce, is highly arch. dependent etc.

File: monty-31/src/utils.rs (hunk 2 above)
Comment (tcoratger) [reply monty-31/src/utils.rs]: Fixed

File: monty-31/src/utils.rs (hunk 3 above)
Comment (tcoratger) [reply monty-31/src/utils.rs]: Yes agreed on all this, fixed!
